### PR TITLE
216 tables

### PR DESCRIPTION
**Pull request by eneo5541**

### CODE REVIEW

File: src/components/ux-table-body/ux-table-body.hbs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<tbody {{#if class}}class="{{.class}}"{{/if}} {{#if id}}id="{{.id}}"{{/if}}>
```
Comment (ivanwills) [inline src/components/ux-table-body/ux-table-body.hbs]: Either you should always use "class" or you should use ".class" (same goes for all other variables) in the above line and lots of other lines in other files you mix and match the styles.

Either the line should be:

``` javascript
<tbody {{#if class}}class="{{class}}"{{/if}} {{#if id}}id="{{id}}"{{/if}}>
```

or

``` javascript
<tbody {{#if .class}}class="{{.class}}"{{/if}} {{#if .id}}id="{{.id}}"{{/if}}>
```

My preference is for the first form (less verbose) but I'll check if there is a reason for the other form.

File: src/components/ux-table/ux-table.steps.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		this.component.column = this.component.container + 'thead td';
+		this.component.row = this.component.container + 'tbody tr';
+
+		callback();
+	});
+
+/*
+ * This is here because the similar assertion inside steps.js does not accept the numElements value provided by the
+ * 'Example' structure in ux-table.feature.
+ */
```
Comment (ivanwills) [inline src/components/ux-table/ux-table.steps.js]: The comment should have the same indentation as the code.

File: src/components/ux-table-body/ux-table-body.hbs (hunk 1 above)
Comment (nathf) [inline src/components/ux-table-body/ux-table-body.hbs]: Agree with @ivanwills but should be the 2nd example.

Reference: http://docs.ractivejs.org/latest/references#current-context-or-this-

**Comment by eneo5541** [comment]
@ivanwills I've refactored the handlebars files for the tables to use the correct contexts.

File: src/components/ux-table-body/ux-table-body.hbs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<tbody {{#if .class}}class="{{.class}}"{{/if}} {{#if .id}}id="{{.id}}"{{/if}}>
```
Comment (scrypto) [inline src/components/ux-table-body/ux-table-body.hbs]: Can you do indentation for this? eg:

``` javascript
<tbody
    {{#if .class}}class="{{.class}}"{{/if}}
    {{#if .id}}id="{{.id}}"{{/if}}>
    {{yield}}
```

File: src/components/ux-table-foot/ux-table-foot.hbs [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<tfoot {{#if .class}}class="{{.class}}"{{/if}} {{#if .id}}id="{{.id}}"{{/if}}>
+	{{yield}}
```
Comment (scrypto) [inline src/components/ux-table-foot/ux-table-foot.hbs]: Please indent the if statements (as per above comments in ux-table-body.hbs)

File: src/components/ux-table-head/ux-table-head.hbs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<thead {{#if .class}}class="{{.class}}"{{/if}} {{#if .id}}id="{{.id}}"{{/if}}>
+	{{yield}}
```
Comment (scrypto) [inline src/components/ux-table-head/ux-table-head.hbs]: Please indent the if statements (as per above comments in ux-table-body.hbs)

File: src/components/ux-table/ux-table.steps.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+		this.component.row = this.component.container + 'tbody tr';
+
+		callback();
+	});
+
+	/*
+	 * This is here because the similar assertion inside steps.js does not accept the numElements value provided by the
+	 * 'Example' structure in ux-table.feature.
+	 */
+	this.Given(/^there are "([^"]*)" "([^"]*)" elements displayed$/, function (numElements, semanticName, callback) {
```
Comment (ivanwills) [inline src/components/ux-table/ux-table.steps.js:21]: You could move this to work.js so others can use

File: src/components/ux-table-row/ux-table-row.hbs [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<tr {{#if .class}}class="{{.class}}"{{/if}} {{#if .id}}id="{{.id}}"{{/if}}>
+	{{>content}}
```
Comment (scrypto) [inline src/components/ux-table-row/ux-table-row.hbs]: Please indent the if statements (as per above comments in ux-table-body.hbs)

File: src/components/ux-table/use-cases/3Rows.json [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+			"class": "header-class",
+			"cell": [
+				{
+					"label": "Table Header"
+				},
+				{
+					"label": "Table Header"
+				},
+				{
+				   "label": "Table Header"
```
Comment (scrypto) [inline src/components/ux-table/use-cases/3Rows.json:16]: Please use tabs instead of spaces

File: src/components/ux-table/use-cases/3Rows.json [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+				},
+				{
+				   "label": "Table Header"
```
Comment (scrypto) [inline src/components/ux-table/use-cases/3Rows.json:19]: Use tab here too

File: src/components/ux-table/use-cases/3Rows.json [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+			"rows": [
+				{
+					"class": "row-class",
+					"cell": [
+						{
+							"label": "Content Goes Here"
+						},
+						{
+						   "class": "cell-class",
+						   "label": "This is longer content Donec id elit non mi porta gravida at eget metus."
```
Comment (scrypto) [inline src/components/ux-table/use-cases/3Rows.json:34]: Use tabs

File: src/components/ux-table/use-cases/3Rows.json [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+				},
+				{
+					"class": "row-class",
+					"cell": [
+						{
+							"label": "Content Goes Here"
+						},
+						{
+						   "class": "cell-class",
+						   "label": "This is longer Content Goes Here Donec id elit non mi porta gravida at eget metus."
```
Comment (scrypto) [inline src/components/ux-table/use-cases/3Rows.json:52]: Use tabs

File: src/components/ux-table/use-cases/3Rows.json [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
```
Comment (scrypto) [inline src/components/ux-table/use-cases/3Rows.json:70]: Tabs, please :)

File: src/components/ux-table/use-cases/4Rows.json [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+			"class": "header-class",
+			"cell": [
+				{
+					"label": "Table Header"
+				},
+				{
+					"label": "Table Header"
+				},
+				{
+				   "label": "Table Header"
```
Comment (scrypto) [inline src/components/ux-table/use-cases/4Rows.json:16]: Tabs

File: src/components/ux-table/use-cases/4Rows.json [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+				},
+				{
+				   "label": "Table Header"
```
Comment (scrypto) [inline src/components/ux-table/use-cases/4Rows.json:19]: Tabs

File: src/components/ux-table/use-cases/4Rows.json [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+			"rows": [
+				{
+					"class": "row-class",
+					"cell": [
+						{
+							"label": "Content Goes Here"
+						},
+						{
+						   "class": "cell-class",
+						   "label": "This is longer content Donec id elit non mi porta gravida at eget metus."
```
Comment (scrypto) [inline src/components/ux-table/use-cases/4Rows.json:34]: Tabs

File: src/components/ux-table/use-cases/4Rows.json [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+				},
+				{
+					"class": "row-class",
+					"cell": [
+						{
+							"label": "Content Goes Here"
+						},
+						{
+						   "class": "cell-class",
+						   "label": "This is longer Content Goes Here Donec id elit non mi porta gravida at eget metus."
```
Comment (scrypto) [inline src/components/ux-table/use-cases/4Rows.json:52]: Tabs

File: src/components/ux-table/use-cases/4Rows.json [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
```
Comment (scrypto) [inline src/components/ux-table/use-cases/4Rows.json:70]: Tabs

File: src/components/ux-table/use-cases/4Rows.json [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
```
Comment (scrypto) [inline src/components/ux-table/use-cases/4Rows.json:88]: Tabs. And please check the rest of the JSON files.

File: src/components/ux-table/use-cases/5Columns.json [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+						{
+							"label": "Content Goes Here"
+						},
+						{
+							"label": "Content Goes Here"
+						},
+                        {
+                          "class": "cell-class",
+                          "label": "This is longer content Donec id elit non mi porta gravida at eget metus."
+                        }
```
Comment (scrypto) [inline src/components/ux-table/use-cases/5Columns.json:48]: Tabs for formatting, please

File: src/components/ux-table/ux-table.feature [hunk 20]
```diff
@@ -0,0 +1,13 @@
+Feature: ux-table test suite
+
+    Scenario Outline: Loading ux-table
+		Given I have loaded component "ux-table" with use case "<mock>"
+		And there will be an element for "table"
+        And there are "<columns>" "column" elements displayed
+        And there are "<rows>" "row" elements displayed
```
Comment (scrypto) [inline src/components/ux-table/ux-table.feature:7]: Need indentation.

File: src/components/ux-table/ux-table.feature [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+		Examples:
+		  | mock      | columns   | rows  |
+		  | 3Rows     | 4         | 3     |
+		  | 4Rows     | 4         | 4     |
+		  | 5Columns  | 5         | 2     |
+		  | 6Columns  | 6         | 2     |
```
Comment (scrypto) [inline src/components/ux-table/ux-table.feature:13]: Please add new line at the end of the file (also check the rest of the commits with the above no-new-line signs)

File: src/components/ux-table/ux-table.hbs [hunk 22]
```diff
@@ -0,0 +1,67 @@
+<table class="ux-table {{class}}"
+       {{#if role}}role="{{role}}"{{/if}}
+        >
+    {{#if caption}}
```
Comment (scrypto) [inline src/components/ux-table/ux-table.hbs]: Please use tabs for indentation instead of spaces (check the rest of the files)

**Comment by nathf** [comment]
We may need to think about this one, what benefit does `tbody`, `thead` etc components add? There was some testing done some time ago on a similar set up and the performance was pretty ordinary.

I would suggest just using native elements for the internals and scrapping all components bar `ux-table`

**Comment by ivanwills** [comment]
@nathf,

One reason for this is the ability to do pure data driven tables with no extra work required. When anyone uses these components they will probably be doing it to quickly get tabular data (probably for quick prototyping) real use should be done with out excessive use of components so may be this should be flagged in the readme that it's probably best due to performance to only use this for prototyping.

**Comment by nathf** [comment]
I think my original comment was confusing what I meant was when will you use `ux-table-body`? Its only ever used internally for the datadriven side. I am just suggesting to swap `ux-table-body` with `tbody` and alike with other not really needed components as they add no real benefit, just additional performance cost of parsing an extra component.

It doesnt really add any benefit to prototyping as everyone already knows the basic HTML tags.

**Comment by ScottMaclure** [comment]
General comment: I think we should focus less on code style and more on overall implementation detail. For strict code style I think this should be automated, to save us time when reviewing.

E.g: http://bikeshed.com/

**Pull request merged**